scheduler/generic_sched_test.go

```diff
@@ -6379,3 +6379,147 @@ func TestPropagateTaskState(t *testing.T) {
+			testutil.WaitForResult(func() (bool, error) {
+				found, err := h.State.EvalByID(ws, followUpEval.ID)
+				if err != nil {
+					return false, err
+				}
+				if found == nil {
+					return false, nil
+				}
+
+				require.Equal(t, nodeStatusUpdateEval.ID, found.PreviousEval)
+				require.Equal(t, "pending", found.Status)
+				require.NotEmpty(t, found.WaitUntil)
+
+				return true, nil
+			}, func(err error) {
+				require.NoError(t, err)
+			})
+
+			// Validate that the ClientStatus updates are part of the plan.
+			require.Len(t, h.Plans[0].NodeAllocation[disconnectedNode.ID], tc.count)
+			// Pending update should have unknown status.
+			for _, nodeAlloc := range h.Plans[0].NodeAllocation[disconnectedNode.ID] {
+				require.Equal(t, nodeAlloc.ClientStatus, structs.AllocClientStatusUnknown)
+			}
+
+			// Simulate that NodeAllocation got processed.
+			testutil.WaitForResult(func() (bool, error) {
+				err = h.State.UpsertAllocs(structs.MsgTypeTestSetup, h.NextIndex(), h.Plans[0].NodeAllocation[disconnectedNode.ID])
+				require.NoError(t, err, "plan.NodeUpdate")
+				return true, nil
+			}, func(err error) {
+				require.NoError(t, err)
+			})
+
+			// Validate that the StateStore Upsert applied the ClientStatus we specified.
+			for _, alloc := range unknownAllocs {
+				alloc, err = h.State.AllocByID(ws, alloc.ID)
+				require.NoError(t, err)
+				require.Equal(t, alloc.ClientStatus, structs.AllocClientStatusUnknown)
+
+				// Allocations have been transitioned to unknown
+				require.Equal(t, structs.AllocDesiredStatusRun, alloc.DesiredStatus)
+				require.Equal(t, structs.AllocClientStatusUnknown, alloc.ClientStatus)
+			}
+
+			// If the reschedule policy indicates they should get rescheduled,
+			// test that they do.
+			if tc.rescheduled {
+
+			}
+		})
+	}
+}
+
+func initNodeAndAllocs(t *testing.T, h *Harness, allocCount int,
+	maxClientDisconnect time.Duration, nodeStatus, clientStatus string) (*structs.Node, *structs.Job, []*structs.Allocation) {
+	// Node, which is ready
+	node := mock.Node()
+	node.Status = nodeStatus
+	require.NoError(t, h.State.UpsertNode(structs.MsgTypeTestSetup, h.NextIndex(), node))
+
+	// Job with allocations and max_client_disconnect
+	job := mock.Job()
+	job.TaskGroups[0].Count = allocCount
+	job.TaskGroups[0].MaxClientDisconnect = &maxClientDisconnect
+	require.NoError(t, h.State.UpsertJob(structs.MsgTypeTestSetup, h.NextIndex(), job))
+
+	allocs := make([]*structs.Allocation, allocCount)
+	for i := 0; i < allocCount; i++ {
+		// Alloc for the running group
+		alloc := mock.Alloc()
+		alloc.Job = job
+		alloc.JobID = job.ID
+		alloc.NodeID = node.ID
+		alloc.Name = fmt.Sprintf("my-job.web[%d]", i)
+		alloc.DesiredStatus = structs.AllocDesiredStatusRun
+		alloc.ClientStatus = clientStatus
+		//if !tc.when.IsZero() {
+		//	alloc.AllocStates = []*structs.AllocState{{
+		//		Field: structs.AllocStateFieldClientStatus,
+		//		Value: structs.AllocClientStatusLost,
+		//		Time:  tc.when,
+		//	}}
+		//}
```

Comment:
What's going on here?

Reply:
I used another test as a basis for this and was unsure if I would need to append `AllocStates` so I left it as a reminder. I should have added a `TODO (derek)` tag with some comments. Sorry for that!

Reply:
Or maybe opening an issue is the right workflow?  As you've seen previously, my habit from `$lastJobs` is to create `TODO` comments. I could use some input so I don't create more cruft for you to have to filter out. Issues feel kind of heavy-handed, but maybe that's preferred? So I'm not sure.

Reply:
I see now in a later comment that the preferred method is to move todos to a new branch. I'll follow that advice.

Reply:
> Or maybe opening an issue is the right workflow? As you've seen previously, my habit from $lastJobs is to create TODO comments. I could use some input so I don't create more cruft for you to have to filter out. Issues feel kind of heavy-handed, but maybe that's preferred? So I'm not sure.

The trouble with TODOs is that they tend to rot in the code base. But this PR is also targeting a feature branch where you could potentially come by and cleanup your TODOs before they land in `main`. So having a TODO here without a corresponding issue is fine so long as it doesn't end up landing in `main` like that. (I would also recommend adding a checklist item to your umbrella issue in team-nomad to make it harder to forget, but that's because I'm super forgetful if I don't use external memory 😀 ). Then once things are getting ready to land in `main` and will live "forever", writing up a quick issue for the TODO will give it a shot at getting done some day.

> I see now in a later comment that the preferred method is to move todos to a new branch. I'll follow that advice.

If you're referring to [this](https://github.com/hashicorp/nomad/pull/12058#discussion_r806147723) comment, what I'm saying there is not to add more stuff onto this already-large PR. Instead, come back and implement it as a new PR so that we can get this one merged and reduce the amount of rebasing you need to do in parallel branches.

Reply:
Removed in next push.